lib/cobra_commander/component_tree.rb

```diff
@@ -8,50 +11,104 @@ def initialize(path)
         {
           name: @name,
           path: @root_path,
+          type: @type,
           ancestry: @ancestry,
-          dependencies: component_dependencies.map(&method(:dep_representation)),
+          dependencies: dependencies.map(&method(:dep_representation)),
         }
       end
 
     private
 
-      def gemfile_path
-        File.join(@root_path, "Gemfile")
+      def type_of_component
+        return "Ruby & JS" if gem? && node?
+        return "Ruby" if gem?
+        return "JS" if node?
       end
 
-      def gemfile_lock_path
-        File.join(@root_path, "Gemfile.lock")
+      def dependencies
+        @deps ||= begin
+          deps = ruby_dependencies + linked_nodes
+          deps.sort_by(&:name)
+        end
+      end
+
+      def ruby_dependencies
+        @ruby_dependencies ||= begin
+          return [] unless gem?
+          bundler_definition.dependencies.select do |dep|
+            dep.source&.is_a_path? && dep.source.path.to_s != "."
+          end
+        end
+      end
+
+      def linked_nodes
+        @nodes ||= begin
+          return [] unless node?
+          json = JSON.parse(File.read(package_json_path))
+          js_format(json["dependencies"])
+        end
+      end
+
+      def js_format(deps)
+        return [] if deps.nil?
+        linked_deps = deps.select { |_, v| v.start_with? "link:" }
+        linked_deps.map do |_, v|
+          relational_path = v.split("link:")[1]
+          dep_name = relational_path.split("/")[-1]
+          OpenStruct.new(name: dep_name, path: relational_path)
+        end
       end
 
       def bundler_definition
         ::Bundler::Definition.build(gemfile_path, gemfile_lock_path, nil)
       end
 
-      def component_dependencies
-        bundler_definition.dependencies.select do |dep|
-          dep.source&.is_a_path? && dep.source.path.to_s != "."
-        end
+      def gem?
+        @gem ||= File.exist?(gemfile_path)
+      end
+
+      def node?
+        @node ||= File.exist?(package_json_path)
+      end
+
+      def package_json_path
+        File.join(@root_path, "package.json")
+      end
+
+      def gemfile_path
+        File.join(@root_path, "Gemfile")
+      end
+
+      def gemfile_lock_path
+        File.join(@root_path, "Gemfile.lock")
       end
 
       def dep_representation(dep)
-        path = File.expand_path(File.join(@root_path, dep.source.path, dep.name))
-        ancestry = @ancestry + [{ name: @name, path: @root_path }]
-        self.class.new(dep.name, path, ancestry).to_h
+        dep_path, dep_name = extract_dep_info(dep)
+        ancestry = @ancestry + [{ name: @name, path: @root_path, type: @type }]
+        self.class.new(dep_name, dep_path, ancestry).to_h
+      end
+
+      def extract_dep_info(dep)
+        relational_path = dep.is_a?(OpenStruct) ? dep.path : "#{dep.source.path}/#{dep.name}"
+        path = File.expand_path(File.join(@root_path, relational_path))
+        [path, dep.name]
       end
     end
   end
```

Comment:
I feel like it would be a good idea to have specialized classes for ruby and javascript components since it seems there is significant logic that they do not share. Then those classes could be called on this one.

Reply:
I first built this out with separate classes. The problem was it turns out the only way to generate a reliable ComponentTree is not to generate separate gem and js trees, because of the use of recursion in `dep_representation`.  There was no way to merge them effectively without a lot of extra looping and equality checking.

In the particular snippet above, the reason a `dep` could be different types is that on the ruby side we rely on `Bundler`.

Reply:
@dmnelson - This is the commit where I remove the separate classes. a5dd89f

Reply:
Ah... I think I see what you mean now :).  Sorry for the confusion. I'll try to extract the logic now.